octoprint_mrbeam/software_update_information.py

```diff
@@ -19,72 +24,197 @@
+                returncode = self.REPO_FILE
+            except IOError as e:
+                _logger.error(
+                    "not even fallback version available: %s",
+                    e,
+                )
+
+        #if file changed, reload update info
+        if newfile:
+            _logger.info("new file => set info")
+
+            # inform SoftwareUpdate Pluging about new config
+            sw_update_plugin = self._plugin._plugin_manager.get_plugin_info(
+                "softwareupdate"
+            ).implementation
+            sw_update_plugin._refresh_configured_checks = True
+            sw_update_plugin._version_cache = dict()
+            sw_update_plugin._version_cache_dirty = True
+        return returncode
+
+@logme(output=True)
 def get_update_information(plugin):
-    result = dict()
-
+    """
+    Gets called from the octoprint.plugin.softwareupdate.check_config Hook from Octoprint
+    Starts a thread to look online for a new config file
+    sets the config for the Octoprint Softwareupdate Plugin with the data from the config file
+    @param plugin: calling plugin
+    @return: the config for the Octoprint embedded softwareupdate Plugin
+    """
     tier = plugin._settings.get(["dev", "software_tier"])
     beamos_tier, beamos_date = plugin._device_info.get_beamos_version()
-    _logger.info("SoftwareUpdate using tier: %s", tier)
-
-    # The increased number of separate virtualenv for iobeam, netconnectd, ledstrips
-    # will increase the "discovery time" to find those package versions.
-    # "map-reduce" method can decrease lookup time by processing them in parallel
-    res = dict(
-        reduce(
-            dict_merge,
-            [
-                _set_info_mrbeam_plugin(plugin, tier, beamos_date),
-                _set_info_mrbeamdoc(plugin, tier),
-                _set_info_netconnectd_plugin(plugin, tier, beamos_date),
-                _set_info_findmymrbeam(plugin, tier),
-                _set_info_mrbeamledstrips(plugin, tier, beamos_date),
-                _set_info_netconnectd_daemon(plugin, tier, beamos_date),
-                _set_info_iobeam(plugin, tier, beamos_date),
-                _set_info_mrb_hw_info(plugin, tier, beamos_date),
-                # _config_octoprint(plugin, tier),
-            ],
-        )
-    )
-    for pack, updt_info in res.items():
-        _logger.debug(
-            "{} targets branch {} using pip {}".format(
-                pack,
-                updt_info.get("branch"),
-                updt_info.get("pip_command", "~/oprint/bin/pip"),
-            )
-        )
-    return res
+    _logger.info("SoftwareUpdate using tier: %s %s", tier, beamos_date)
+    global _softwareupdate_handler
+    if _softwareupdate_handler is None:
+        _softwareupdate_handler = MrBeamSoftwareupdateHandler(plugin)
+        _softwareupdate_handler.load_update_file_from_cloud()
+
+    return _set_info_from_file(plugin, tier, beamos_date, _softwareupdate_handler)
```

Comment:
please do not use singletons. You can just create a new instance, there is no benefit to have a singleton here.

Reply:
but then it can run multiple processes that will fetch the file from the server, this will make sure that only one process is running that checks the file

Reply:
Please use a lock if you want to manage threaded access.

Reply:
is it done correlty now?